File: packages/virage-core/src/quality/report.ts
/**
 * Report formatters for QualityReport.
 * Three formats: JSON (machine), Markdown (PR comments), Console (colored table).
 */

import type { QualityReport } from "./interfaces.js";

// ─── JSON ─────────────────────────────────────────────────────────────────────

export function formatJson(report: QualityReport): string {
  return JSON.stringify(report, null, 2);
}

// ─── Markdown ─────────────────────────────────────────────────────────────────

export function formatMarkdown(report: QualityReport): string {
  const statusEmoji = report.status === "PASS" ? "✅" : "❌";
  const scorePercent = (report.overallScore * 100).toFixed(1);

  const lines: string[] = [
    `## ${statusEmoji} Virage Quality Report`,
    ``,
    `**Overall Score:** ${scorePercent}% — **${report.status}**  `,
    `**Timestamp:** ${report.timestamp}  `,
    `**Sample Size:** ${report.sampleSize} chunks | **Top-K:** ${report.topK}`,
    ``,
  ];

  // Must-pass gates
  const failedGates = report.mustPassGates.filter((g) => !g.passed);
  if (failedGates.length > 0) {
    lines.push(`### ⚠️ Must-Pass Failures`);
    lines.push(``);
    lines.push(`| Metric | Value | Threshold | Status |`);
    lines.push(`|--------|-------|-----------|--------|`);
    for (const gate of failedGates) {
      lines.push(
        `| ${gate.metricName} | ${gate.value.toFixed(3)} | >${gate.threshold} | ❌ FAIL |`,
      );
    }
    lines.push(``);
  }

  // Component scores
  lines.push(`### Component Scores`);
  lines.push(``);
  lines.push(`| Component | Score | Status |`);
  lines.push(`|-----------|-------|--------|`);
  for (const comp of report.components) {
    if (comp.skipped) {
      lines.push(`| ${comp.label} | — | ⏭ skipped |`);
    } else {
      const scoreStr = (comp.score * 100).toFixed(1) + "%";
      const emoji = comp.score >= 0.7 ? "🟢" : comp.score >= 0.5 ? "🟡" : "🔴";
      lines.push(`| ${comp.label} | ${scoreStr} | ${emoji} |`);
    }
  }
  lines.push(``);

  // RAGBench results
  if (report.ragBench) {
    const rb = report.ragBench;
    lines.push(`### RAGBench Retrieval Evaluation`);
    lines.push(``);
    lines.push(
      `**Source:** \`${rb.datasetSource}\` — **Queries evaluated:** ${rb.queriesEvaluated} | **K:** ${rb.topK}`,
    );
    lines.push(``);
    lines.push(`| Metric | Score |`);
    lines.push(`|--------|-------|`);
    lines.push(`| MRR@${rb.topK} | ${(rb.mrrAtK * 100).toFixed(2)}% |`);
    lines.push(`| NDCG@${rb.topK} | ${(rb.ndcgAtK * 100).toFixed(2)}% |`);
    lines.push(`| Recall@${rb.topK} | ${(rb.recallAtK * 100).toFixed(2)}% |`);
    lines.push(
      `| Precision@${rb.topK} | ${(rb.precisionAtK * 100).toFixed(2)}% |`,
    );
    lines.push(`| HitRate@${rb.topK} | ${(rb.hitRateAtK * 100).toFixed(2)}% |`);
    lines.push(``);
  }

  // Metric details (collapsed)
  lines.push(`<details>`);
  lines.push(`<summary>Metric Details</summary>`);
  lines.push(``);
  for (const comp of report.components) {
    if (comp.skipped) continue;
    lines.push(`#### ${comp.label}`);
    lines.push(``);
    lines.push(`| Metric | Raw | Normalized | Weight | Status |`);
    lines.push(`|--------|-----|------------|--------|--------|`);
    for (const m of comp.metrics) {
      if (m.skipped) {
        lines.push(
          `| ${m.name} | — | — | ${m.weight} | ⏭ ${m.skipReason ?? "skipped"} |`,
        );
      } else {
        const mustPass = m.mustPass
          ? m.mustPassPassed
            ? " ✅"
            : " ❌ MUST-PASS FAIL"
          : "";
        lines.push(
          `| ${m.name} | ${m.rawValue.toFixed(3)} | ${(m.normalizedValue * 100).toFixed(1)}% | ${m.weight} | —${mustPass} |`,
        );
      }
    }
    lines.push(``);
  }
  lines.push(`</details>`);

  return lines.join("\n");
}

// ─── Console ──────────────────────────────────────────────────────────────────

const ansi = {
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  red: "\x1b[31m",
  cyan: "\x1b[36m",
  gray: "\x1b[90m",
};

function scoreColor(score: number): string {
  if (score >= 0.7) return ansi.green;
  if (score >= 0.5) return ansi.yellow;
  return ansi.red;
}

function bar(score: number, width = 20): string {
  const filled = Math.round(score * width);
  return "█".repeat(filled) + "░".repeat(width - filled);
}

export function formatConsole(report: QualityReport): string {
  const statusColor = report.status === "PASS" ? ansi.green : ansi.red;
  const scorePercent = (report.overallScore * 100).toFixed(1);

  const lines: string[] = [
    "",
    `${ansi.bold}╔════════════════════════════════════════════════════╗${ansi.reset}`,
    `${ansi.bold}║            Virage Quality Report                   ║${ansi.reset}`,
    `${ansi.bold}╚════════════════════════════════════════════════════╝${ansi.reset}`,
    "",
    `  Status      : ${statusColor}${ansi.bold}${report.status}${ansi.reset}`,
    `  Score       : ${scoreColor(report.overallScore)}${scorePercent}%${ansi.reset}  ${bar(report.overallScore)}`,
    `  Sample size : ${report.sampleSize} chunks  |  Top-K: ${report.topK}`,
    `  Duration    : ${report.durationMs}ms`,
    "",
  ];

  // Must-pass failures
  const failedGates = report.mustPassGates.filter((g) => !g.passed);
  if (failedGates.length > 0) {
    lines.push(`  ${ansi.red}${ansi.bold}⚠ Must-Pass Failures${ansi.reset}`);
    for (const gate of failedGates) {
      lines.push(
        `  ${ansi.red}✗ ${gate.metricName}: ${gate.value.toFixed(3)} (threshold: >${gate.threshold})${ansi.reset}`,
      );
    }
    lines.push("");
  }

  // Component table
  lines.push(
    `  ${ansi.bold}${"COMPONENT".padEnd(22)} ${"SCORE".padEnd(8)} BAR${ansi.reset}`,
  );
  lines.push(`  ${"─".repeat(52)}`);

  for (const comp of report.components) {
    if (comp.skipped) {
      lines.push(`  ${comp.label.padEnd(22)} ${ansi.dim}skipped${ansi.reset}`);
      continue;
    }
    const scoreStr = `${(comp.score * 100).toFixed(1)}%`;
    const color = scoreColor(comp.score);
    lines.push(
      `  ${comp.label.padEnd(22)} ${color}${scoreStr.padEnd(8)}${ansi.reset} ${bar(comp.score, 18)}`,
    );

    // Show must-pass metrics inline
    for (const m of comp.metrics) {
      if (m.mustPass && !m.skipped) {
        const icon = m.mustPassPassed ? `${ansi.green}✓` : `${ansi.red}✗`;
        lines.push(
          `  ${ansi.dim}  └─ ${m.name}: ${m.rawValue.toFixed(3)}${ansi.reset} ${icon}${ansi.reset}`,
        );
      }
    }
  }

  lines.push(`  ${"─".repeat(52)}`);
  lines.push(
    `  ${"OVERALL".padEnd(22)} ${scoreColor(report.overallScore)}${scorePercent}%${ansi.reset.padEnd(7 - scorePercent.length)} ${bar(report.overallScore, 18)}`,
  );
  lines.push("");

  // RAGBench section
  if (report.ragBench) {
    const rb = report.ragBench;
    lines.push(
      `  ${ansi.bold}RAGBench — ${rb.datasetSource} (${rb.queriesEvaluated} queries, K=${rb.topK})${ansi.reset}`,
    );
    lines.push(`  ${"─".repeat(52)}`);
    const rbRows: Array<[string, number]> = [
      [`MRR@${rb.topK}`, rb.mrrAtK],
      [`NDCG@${rb.topK}`, rb.ndcgAtK],
      [`Recall@${rb.topK}`, rb.recallAtK],
      [`Precision@${rb.topK}`, rb.precisionAtK],
      [`HitRate@${rb.topK}`, rb.hitRateAtK],
    ];
    for (const [name, val] of rbRows) {
      const valStr = `${(val * 100).toFixed(1)}%`;
      lines.push(
        `  ${name.padEnd(22)} ${scoreColor(val)}${valStr.padEnd(8)}${ansi.reset} ${bar(val, 18)}`,
      );
    }
    lines.push("");
  }

  return lines.join("\n");
}

File: crates/virage-engine/src/quality/report.rs
//! Report formatters for `QualityReport`. Three formats: JSON (machine), Markdown (PR
//! comments), Console (colored table). Ported from `dist/quality/report.js` — IR-038.

use super::QualityReport;

pub fn format_json(report: &QualityReport) -> String {
    serde_json::to_string_pretty(report).unwrap_or_default()
}

pub fn format_markdown(report: &QualityReport) -> String {
    let status_emoji = if report.status.is_pass() {
        "✅"
    } else {
        "❌"
    };
    let score_percent = report.overall_score * 100.0;
    let mut lines = vec![
        format!("## {status_emoji} Virage Quality Report"),
        String::new(),
        format!(
            "**Overall Score:** {score_percent:.1}% — **{}**  ",
            report.status
        ),
        format!("**Timestamp:** {}  ", report.timestamp),
        format!(
            "**Sample Size:** {} chunks | **Top-K:** {}",
            report.sample_size, report.top_k
        ),
        String::new(),
    ];

    let failed_gates: Vec<_> = report
        .must_pass_gates
        .iter()
        .filter(|g| !g.passed)
        .collect();
    if !failed_gates.is_empty() {
        lines.push("### ⚠️ Must-Pass Failures".to_string());
        lines.push(String::new());
        lines.push("| Metric | Value | Threshold | Status |".to_string());
        lines.push("|--------|-------|-----------|--------|".to_string());
        for gate in &failed_gates {
            lines.push(format!(
                "| {} | {:.3} | >{} | ❌ FAIL |",
                gate.metric_name, gate.value, gate.threshold
            ));
        }
        lines.push(String::new());
    }

    lines.push("### Component Scores".to_string());
    lines.push(String::new());
    lines.push("| Component | Score | Status |".to_string());
    lines.push("|-----------|-------|--------|".to_string());
    for comp in &report.components {
        if comp.skipped {
            lines.push(format!("| {} | — | ⏭ skipped |", comp.label));
        } else {
            let score_str = format!("{:.1}%", comp.score * 100.0);
            let emoji = if comp.score >= 0.7 {
                "🟢"
            } else if comp.score >= 0.5 {
                "🟡"
            } else {
                "🔴"
            };
            lines.push(format!("| {} | {score_str} | {emoji} |", comp.label));
        }
    }
    lines.push(String::new());

    lines.push("<details>".to_string());
    lines.push("<summary>Metric Details</summary>".to_string());
    lines.push(String::new());
    for comp in &report.components {
        if comp.skipped {
            continue;
        }
        lines.push(format!("#### {}", comp.label));
        lines.push(String::new());
        lines.push("| Metric | Raw | Normalized | Weight | Status |".to_string());
        lines.push("|--------|-----|------------|--------|--------|".to_string());
        for m in &comp.metrics {
            if m.skipped {
                lines.push(format!(
                    "| {} | — | — | {} | ⏭ {} |",
                    m.name,
                    m.weight,
                    m.skip_reason.as_deref().unwrap_or("skipped")
                ));
            } else {
                let must_pass = if m.must_pass {
                    if m.must_pass_passed == Some(true) {
                        " ✅"
                    } else {
                        " ❌ MUST-PASS FAIL"
                    }
                } else {
                    ""
                };
                lines.push(format!(
                    "| {} | {:.3} | {:.1}% | {} | —{must_pass} |",
                    m.name,
                    m.raw_value,
                    m.normalized_value * 100.0,
                    m.weight
                ));
            }
        }
        lines.push(String::new());
    }
    lines.push("</details>".to_string());
    lines.join("\n")
}

mod ansi {
    pub const RESET: &str = "\x1b[0m";
    pub const BOLD: &str = "\x1b[1m";
    pub const DIM: &str = "\x1b[2m";
    pub const GREEN: &str = "\x1b[32m";
    pub const YELLOW: &str = "\x1b[33m";
    pub const RED: &str = "\x1b[31m";
}

fn score_color(score: f64) -> &'static str {
    if score >= 0.7 {
        ansi::GREEN
    } else if score >= 0.5 {
        ansi::YELLOW
    } else {
        ansi::RED
    }
}

fn bar(score: f64, width: usize) -> String {
    let filled = (score * width as f64).round().max(0.0) as usize;
    let filled = filled.min(width);
    format!("{}{}", "█".repeat(filled), "░".repeat(width - filled))
}

pub fn format_console(report: &QualityReport) -> String {
    let status_color = if report.status.is_pass() {
        ansi::GREEN
    } else {
        ansi::RED
    };
    let score_percent = report.overall_score * 100.0;
    let reset = ansi::RESET;
    let bold = ansi::BOLD;
    let dim = ansi::DIM;

    let mut lines = vec![
        String::new(),
        format!("{bold}╔════════════════════════════════════════════════════╗{reset}"),
        format!("{bold}║            Virage Quality Report                   ║{reset}"),
        format!("{bold}╚════════════════════════════════════════════════════╝{reset}"),
        String::new(),
        format!(
            "  Status      : {status_color}{bold}{}{reset}",
            report.status
        ),
        format!(
            "  Score       : {}{score_percent:.1}%{reset}  {}",
            score_color(report.overall_score),
            bar(report.overall_score, 20)
        ),
        format!(
            "  Sample size : {} chunks  |  Top-K: {}",
            report.sample_size, report.top_k
        ),
        format!("  Duration    : {}ms", report.duration_ms),
        String::new(),
    ];

    let failed_gates: Vec<_> = report
        .must_pass_gates
        .iter()
        .filter(|g| !g.passed)
        .collect();
    if !failed_gates.is_empty() {
        lines.push(format!("  {}{bold}⚠ Must-Pass Failures{reset}", ansi::RED));
        for gate in &failed_gates {
            lines.push(format!(
                "  {}✗ {}: {:.3} (threshold: >{}){reset}",
                ansi::RED,
                gate.metric_name,
                gate.value,
                gate.threshold
            ));
        }
        lines.push(String::new());
    }

    lines.push(format!(
        "  {bold}{:<22} {:<8} BAR{reset}",
        "COMPONENT", "SCORE"
    ));
    lines.push(format!("  {}", "─".repeat(52)));
    for comp in &report.components {
        if comp.skipped {
            lines.push(format!("  {:<22} {dim}skipped{reset}", comp.label));
            continue;
        }
        let score_str = format!("{:.1}%", comp.score * 100.0);
        let color = score_color(comp.score);
        lines.push(format!(
            "  {:<22} {color}{:<8}{reset} {}",
            comp.label,
            score_str,
            bar(comp.score, 18)
        ));
        for m in &comp.metrics {
            if m.must_pass && !m.skipped {
                let icon = if m.must_pass_passed == Some(true) {
                    format!("{}✓", ansi::GREEN)
                } else {
                    format!("{}✗", ansi::RED)
                };
                lines.push(format!(
                    "  {dim}  └─ {}: {:.3}{reset} {icon}{reset}",
                    m.name, m.raw_value
                ));
            }
        }
    }
    lines.push(format!("  {}", "─".repeat(52)));
    lines.push(format!(
        "  {:<22} {}{score_percent:.1}%{reset} {}",
        "OVERALL",
        score_color(report.overall_score),
        bar(report.overall_score, 18)
    ));
    lines.push(String::new());
    lines.join("\n")
}
